core/arch/arm/crypto/aes-gcm-ce.c

```diff
@@ -78,33 +92,84 @@ TEE_Result internal_aes_gcm_expand_enc_key(const void *key, size_t key_len,
 }
 
 #ifdef ARM64
-void
-internal_aes_gcm_update_payload_blocks(struct internal_aes_gcm_state *state,
-				       const struct internal_aes_gcm_key *ek,
-				       TEE_OperationMode mode, const void *src,
-				       size_t num_blocks, void *dst)
+static void update_payload_2block(struct internal_aes_gcm_state *state,
+				  const struct internal_aes_gcm_key *ek,
+				  TEE_OperationMode mode, const void *src,
+				  size_t num_blocks, void *dst)
 {
 	uint32_t vfp_state;
 	uint64_t dg[2];
-	uint64_t ctr[2];
+
+	assert(num_blocks && !(num_blocks % 2));
 
 	get_be_block(dg, state->hash_state);
-	get_be_block(ctr, state->ctr);
 
 	vfp_state = thread_kernel_enable_vfp();
 
-	pmull_gcm_load_round_keys(ek->data, ek->rounds);
-
-	if (mode == TEE_MODE_ENCRYPT)
+	if (mode == TEE_MODE_ENCRYPT) {
+		uint8_t ks[sizeof(state->buf_cryp) * 2];
+
+		/*
+		 * ks holds the encrypted counters of the next two block.
+		 * pmull_gcm_encrypt() uses this to encrypt the first two
+		 * blocks. When pmull_gcm_encrypt() returns is ks updated
+		 * with the encrypted counters of the next two blocks. As
+		 * we're only keeping one of these blocks we throw away
+		 * block number two consequently decreases the counter by
+		 * one.
+		 */
+		memcpy(ks, state->buf_cryp, sizeof(state->buf_cryp));
+
+		pmull_gcm_load_round_keys(ek->data, ek->rounds);
+		pmull_gcm_encrypt_block(ks + sizeof(state->buf_cryp),
+					(uint8_t *)state->ctr, ek->rounds);
+		internal_aes_gcm_inc_ctr(state);
 		pmull_gcm_encrypt(num_blocks, dg, dst, src, &state->ghash_key,
-				  ctr, ek->rounds, state->buf_cryp);
-	else
+				  state->ctr, NULL, ek->rounds, ks);
+		memcpy(state->buf_cryp, ks, TEE_AES_BLOCK_SIZE);
+		internal_aes_gcm_dec_ctr(state);
+	} else {
 		pmull_gcm_decrypt(num_blocks, dg, dst, src, &state->ghash_key,
-				  ctr, ek->rounds);
+				  state->ctr, ek->data, ek->rounds);
+	}
 
 	thread_kernel_disable_vfp(vfp_state);
 
-	put_be_block(state->ctr, ctr);
 	put_be_block(state->hash_state, dg);
 }
+
+/* Overriding the __weak function */
+void
+internal_aes_gcm_update_payload_blocks(struct internal_aes_gcm_state *state,
+				       const struct internal_aes_gcm_key *ek,
+				       TEE_OperationMode mode, const void *src,
+				       size_t num_blocks, void *dst)
+{
+	size_t nb = num_blocks / 2 * 2;
```

Comment:
sure compiler will preserved this? maybe `nb = ROUNDDOWN(num_block, 2);`

Reply:
Yeah, it works but you're code looks better.